design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+the values for the fields will be verified against the schema. 
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+If the user specifies `AVRO` encoding in their `CREATE` statement, we will register the topic with 
+the schema registry proactively if the topic did not previously exist. Note that this will not 
+change how we handle `CREATE` statements for topics that already exist (today we do not validate
+that the schema supplied matches the schema in Schema Registry).
+
+### ROWKEY and ROWTIME Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT VALUES`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "timestamp": 1, "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. 
+
+The column schema will always include the implicit `ROWKEY` and `ROWTIME` fields, and the user can
+choose to specify or ignore these fields. To prevent the inconsistencies described above, though,
+we will automatically enforce consistency whenever the key property is set in the `WITH` clause.
+
+* If the user sets a `KEY` property, they do not need to supply `ROWKEY` in their `INSERT INTO`
+statement (e.g. `INSERT INTO foo (id, foo) VALUES ("key", 123);`). In this case, the kafka message
+produced will automatically populate `ROWKEY` to the value of `id`: 
+  ```json
+  {"key": "key", "timestamp": 1, "value": {"id": "key", "foo": 123}}
+  ```
+* Analogous to the above, if the user provides the key but not `ROWKEY`, it will be populated for 
+them.
+* If the user sets a `KEY` property _and_ they supply `ROWKEY` in their `INSERT INTO` statement, 
+KSQL will enforce that the value of `ROWKEY` and the value of the key column are identical.
+  ```sql
+  INSERT INTO foo (ROWKEY, ID, VALUE) VALUES ("THIS WILL FAIL", "key", 123);
+  ```
+* If the user does not set a `KEY` property, and they are inserting into a stream, they may choose
+to omit `ROWKEY` altogether. This will result in a `null` key.
+* If the user does not set a `KEY` property, and they are inserting into a table, they _must_
+have a `ROWKEY` value in their `INSERT INTO` statement.
+
+If a value for `ROWTIME` is supplied, then that will be populated into the kafka record,
+otherwise broker time is used. If a timestamp extractor is also supplied, the `ROWTIME` value will
```

Comment:
> If a value for `ROWTIME` is supplied, then that will be populated into the kafka record

For clarification, by populated into the kafka record you mean set as the record's timestamp?

> If a timestamp extractor is also supplied, the `ROWTIME` value will still be sent.

nit: this is technically not necessary. Any query that consumes the stream/table will do this automatically.

> broker time is used

Why broker time? Seems like (especially for testing purposes) it would be most intuitive to use the local timem.

Reply:
1. yes
2. yes, but I wanted to call that out since it's harder to figure out if there is an extractor and not to send it vs. just sending it and letting it be overridden
3. I think this makes sense. I will do that instead